xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-default/src/main/java/org/xwiki/mentions/internal/async/MapBasedLinkedBlockingQueue.java

```diff
@@ -0,0 +1,364 @@
+        }
+    }
+
+    /**
+     * Clear the internal map and reset the highest key to 0.
+     */
+    private void clearInternalMap()
+    {
+        this.internalMap.clear();
+        this.highestKey.set(0);
+    }
+
+    @Override
+    public void put(T t) throws InterruptedException
+    {
+        Long key = this.internalPut(t);
+        this.internalQueue.put(Pair.of(key, t));
+    }
+
+    @Override
+    public boolean offer(T t, long l, TimeUnit timeUnit) throws InterruptedException
+    {
+        Long key = this.internalPut(t);
+        return this.internalQueue.offer(Pair.of(key, t), l, timeUnit);
+    }
+
+    @Override
+    public boolean offer(T t)
+    {
+        Long key = this.internalPut(t);
+        return this.internalQueue.offer(Pair.of(key, t));
+    }
+
+    @Override
+    public T take() throws InterruptedException
+    {
+        Pair<Long, T> result = this.internalQueue.take();
+        this.internalRemove(result);
+        return result.getValue();
+    }
+
+    @Override
+    public T poll(long l, TimeUnit timeUnit) throws InterruptedException
+    {
+        Pair<Long, T> result = this.internalQueue.poll(l, timeUnit);
+        this.internalRemove(result);
+        return result.getValue();
+    }
+
+    @Override
+    public int remainingCapacity()
+    {
+        return this.internalQueue.remainingCapacity();
+    }
+
+    @Override
+    public T poll()
+    {
+        Pair<Long, T> result = this.internalQueue.poll();
+        this.internalRemove(result);
+        return result.getValue();
+    }
+
+    @Override
+    public T element()
+    {
+        return this.internalQueue.element().getValue();
+    }
+
+    @Override
+    public T peek()
+    {
+        Pair<Long, T> result = this.internalQueue.peek();
+        return result.getValue();
+    }
+
+    @Override
+    public boolean remove(Object o)
+    {
+        boolean result = this.internalQueue.remove(o);
+        this.internalRemove(o);
+        return result;
+    }
+
+    @Override
+    public boolean containsAll(Collection<?> collection)
+    {
+        for (Object o : collection) {
+            if (!this.contains(o)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    @Override
+    public int size()
+    {
+        return this.internalQueue.size();
+    }
+
+    @Override
+    public boolean isEmpty()
+    {
+        return this.internalQueue.isEmpty();
+    }
+
+    @Override
+    public boolean contains(Object o)
+    {
+        // We check on the map since most likely the given object is not a Pair but a T.
+        return this.internalMap.containsValue(o);
+    }
+
+    @Override
+    public Iterator<T> iterator()
+    {
+        return this.internalMap.entrySet().stream().sorted(new DefaultComparator()).map(Map.Entry::getValue).iterator();
+    }
+
+    @Override
+    public Object[] toArray()
+    {
+        return this.internalMap.values().toArray();
```

Comment:
Same question, why use the map here ? The map should be used only to add/remove elements IMO.

Reply:
Same answer, because of types. Here it's a bit less of an issue since it returns Object[] but I don't think the users would expect to have Pair in the result instead of T.

Reply:
Same answer that above, you can easily use stream mapping and it would be much faster than the map that might be slow (definitely is when the map is on filesystem).